### PR TITLE
SYNCOPE-1511: Provide Audit APIs

### PR DESCRIPTION
**Pull request by mmoayyed**
This pull request demonstrates the basic outline of introduced components and the design, modeled and inspired by existing functionality/components in Syncope. This is still very much a WIP and will be updated continuously.

A brief outline of the changes are:

1. AuditService as the main REST API interface, presenting search ops
2. AuditLogic in charge of executing search ops, etc.
3. AuditDAO and AuditDataBinder, in charge of handling the JPA layer and data bindings.
4. AuditTO, AuditEntry and AuditEntryImpl as carrier/transfer objects between the data layer, core and APIs.

### CODE REVIEW

File: common/lib/src/main/java/org/apache/syncope/common/lib/to/AuditTO.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.XmlType;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+
+@XmlRootElement(name = "audit")
+@XmlType
+public class AuditTO extends BaseBean implements EntityTO {
```
Comment (ilgrosso) [inline common/lib/src/main/java/org/apache/syncope/common/lib/to/AuditTO.java]: This should be renamed after `AuditEntryTO`.
Also, I don't see reasons for implementing `EntityTO`, which should be reserved to transfer objects bound to JPA entities.

File: core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.apache.syncope.core.persistence.api.dao.search.OrderByClause;
+import org.apache.syncope.core.persistence.api.dao.search.SearchCond;
+import org.apache.syncope.core.persistence.api.entity.AuditEntry;
+
+import java.util.List;
+import java.util.Set;
+
+public interface AuditDAO<T extends AuditEntry> {
+    AuditEntry findByUsername(String who);
```
Comment (ilgrosso) [inline core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java]: Searching for Audit entries should not be limited to users: maybe `findByEntityKey()` would make sense as we are planning to lookup entries related to Users, Groups and AnyObjects (for the moment) but possibly also for Resource and Connnector Instances in the near future.

File: core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+package org.apache.syncope.core.persistence.api.dao;
+
+import org.apache.syncope.core.persistence.api.dao.search.OrderByClause;
+import org.apache.syncope.core.persistence.api.dao.search.SearchCond;
+import org.apache.syncope.core.persistence.api.entity.AuditEntry;
+
+import java.util.List;
+import java.util.Set;
+
+public interface AuditDAO<T extends AuditEntry> {
```
Comment (ilgrosso) [inline core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java]: As Audit entries are not managed by Persistence Layer, but by `AuditManager` I would rather move all classes to Provisioning Layer, next to it.

File: fit/core-reference/pom.xml [hunk 4]
```diff
@@ -338,7 +338,7 @@ under the License.
             <deployable>
               <location>${project.build.directory}/${project.build.finalName}</location>
               <pingURL>http://localhost:${cargo.servlet.port}/syncope/cacheStats.jsp</pingURL>
-              <pingTimeout>60000</pingTimeout>
+              <pingTimeout>180000</pingTimeout>
```
Comment (ilgrosso) [inline fit/core-reference/pom.xml]: Why is such change needed?

Comment (mmoayyed) [reply fit/core-reference/pom.xml]: I can pull this back if necessary but this is a "backport" from the master where it allows folks with less-capable machines to run fits with more forgiving timeouts. Depending on availability of system resources (specially I think after one's run fits a few times), the tests seems to start timing out under the old timeout value before Syncope has had a chance to bring itself up.

File: common/lib/src/main/java/org/apache/syncope/common/lib/to/AuditTO.java (hunk 1 above)
Comment (mmoayyed) [reply common/lib/src/main/java/org/apache/syncope/common/lib/to/AuditTO.java]: This seems required by the logic API. All logic implementations require the type `<T extends EntityTO>`. Should we remove the concept of AuditLogic or is there an alternative to the internal API?

Comment (ilgrosso) [reply common/lib/src/main/java/org/apache/syncope/common/lib/to/AuditTO.java]: mmm you're right, actually; let me resolve this item...

File: fit/core-reference/pom.xml (hunk 4 above)
Comment (ilgrosso) [reply fit/core-reference/pom.xml]: Ok; after all it does not provoke any bad consequence.

File: core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java (hunk 3 above)
Comment (mmoayyed) [reply core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java]: API work almost done. Going to move classes to the persistence module and clean up.

Comment (ilgrosso) [reply core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java]: Agree, let's first ensure things work fine - I am also not completely sure of my assert above, actually...

File: core/logic/src/main/java/org/apache/syncope/core/logic/AuditLogic.java [hunk 5]
```diff
@@ -41,60 +35,31 @@
...
-            effectiveSearchCond, page, size, orderBy);
-        List<AuditTO> result = matching.stream().
-            map(binder::getAuditTO).
+        List<AuditEntry> matching = auditDAO.findByEntityKey(key, page, size, orderByClauses);
+        List<AuditEntryTO> results = matching.stream().
+            map(audit -> binder.returnAuditTO(binder.getAuditTO(audit))).
             collect(Collectors.toList());
-
-        return Pair.of(count, result);
+        return Pair.of(results.size(), results);
```
Comment (ilgrosso) [inline core/logic/src/main/java/org/apache/syncope/core/logic/AuditLogic.java]: why `results.size()` and not (the dismissed) `count()`?

File: core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java [hunk 6]
```diff
@@ -19,50 +19,88 @@
...
+                 queryString += " ORDER BY " + orderByClauses.
+                    stream().
+                    map(orderBy -> orderBy.getField() + " " + orderBy.getDirection().name()).
+                    collect(Collectors.joining(","));
+            }
+            JdbcTemplate template = getJdbcTemplate();
+            template.setMaxRows(itemsPerPage);
+            template.setFetchSize(itemsPerPage * (page <= 0 ? 0 : page - 1));
+            return template.query(queryString, new AuditEntryRowMapper(key));
+        } catch (SyncopeClientException e) {
```
Comment (ilgrosso) [inline core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java]: What is supposed to throw `SyncopeClientException` here?

File: core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java [hunk 7]
```diff
@@ -19,50 +19,88 @@
...
+            throw e;
+        } catch (Exception e) {
+            LOG.error(e.getMessage(), e);
```
Comment (ilgrosso) [inline core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java]: Can you add some context to the log message?

File: core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java [hunk 8]
```diff
@@ -19,50 +19,88 @@
...
+            if (!orderByClauses.isEmpty()) {
+                 queryString += " ORDER BY " + orderByClauses.
+                    stream().
+                    map(orderBy -> orderBy.getField() + " " + orderBy.getDirection().name()).
+                    collect(Collectors.joining(","));
+            }
+            JdbcTemplate template = getJdbcTemplate();
+            template.setMaxRows(itemsPerPage);
+            template.setFetchSize(itemsPerPage * (page <= 0 ? 0 : page - 1));
+            return template.query(queryString, new AuditEntryRowMapper(key));
```
Comment (ilgrosso) [inline core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java]: Why not lambda here?

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java [hunk 9]
```diff
@@ -1,19 +1,60 @@
...
-        return null;
+    public AuditEntryTO getAuditTO(final AuditEntry auditEntry) {
+        AuditEntryTO auditTO = new AuditEntryTO();
+        auditTO.setKey(auditEntry.getKey());
+        auditTO.setWho(auditEntry.getWho());
+        auditTO.setDate(auditEntry.getDate());
+        auditTO.setThrowable(auditEntry.getThrowable());
+        auditTO.setLoggerName(auditEntry.getLogger().toLoggerName());
+
+        if (StringUtils.isNotBlank(auditEntry.getLogger().getSubcategory())) {
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: What's the purpose of such `if`? If `auditEntry.getLogger().getSubcategory()` is `null`, then no troubles in setting the newly created `audiTO`'s `subCategory` as `null` too.

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java [hunk 10]
```diff
@@ -1,19 +1,60 @@
...
+            auditTO.setSubCategory(auditEntry.getLogger().getSubcategory());
+        }
+        if (StringUtils.isNotBlank(auditEntry.getLogger().getEvent())) {
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: Same as above.

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java [hunk 11]
```diff
@@ -1,19 +1,60 @@
...
+                auditEntry.getOutput(), ToStringStyle.JSON_STYLE));
+        }
+        
+        return auditTO;
     }
 
     @Override
-    public AuditTO getAuditTO(final AuditEntry application) {
-        return null;
+    public AuditEntryTO returnAuditTO(final AuditEntryTO user) {
```
Comment (ilgrosso) [inline core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: `user`??!?

File: core/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/AbstractAnyService.java [hunk 12]
```diff
@@ -128,7 +128,7 @@ public TO read(final String key) {
         SearchCond searchCond = StringUtils.isBlank(anyQuery.getFiql())
                 ? null
                 : getSearchCond(anyQuery.getFiql(), realm);
-
+
```
Comment (ilgrosso) [inline core/rest-cxf/src/main/java/org/apache/syncope/core/rest/cxf/service/AbstractAnyService.java]: I guess you can easily revert such a change.

File: fit/core-reference/src/test/java/org/apache/syncope/fit/core/AuditITCase.java [hunk 13]
```diff
@@ -46,35 +51,57 @@ public static UserTO getSampleUserTO(final String email) {
         return userTO;
     }
 
-    @BeforeEach
-    public void setup() {
-        /**
-         * Execute operations to activate audits.
-         */
+    private static Optional<UserTO> findUser(final String username) {
         PagedResult<UserTO> matching = userService.search(
```
Comment (ilgrosso) [inline fit/core-reference/src/test/java/org/apache/syncope/fit/core/AuditITCase.java]: Why not simply `userService.read(username)` to replace the whole `findUser()` method?

File: core/logic/src/main/java/org/apache/syncope/core/logic/AuditLogic.java (hunk 5 above)
Comment (mmoayyed) [reply core/logic/src/main/java/org/apache/syncope/core/logic/AuditLogic.java]: You're right! I missed the fact count() does not take into account sizing and paging flags and so `count()` might in fact return a different result than `result.size()`. I'll correct this. Thanks!

File: core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java (hunk 6 above)
Comment (mmoayyed) [reply core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java]: Cleaned up.

File: core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java (hunk 7 above)
Comment (mmoayyed) [reply core/persistence-jpa/src/main/java/org/apache/syncope/core/persistence/jpa/dao/JPAAuditDAO.java]: Updated error message to be clearer

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java (hunk 9 above)
Comment (mmoayyed) [reply core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: Cleaned up.

File: core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java (hunk 11 above)
Comment (mmoayyed) [reply core/provisioning-java/src/main/java/org/apache/syncope/core/provisioning/java/data/AuditDataBinderImpl.java]: Cleaned up.

File: fit/core-reference/src/test/java/org/apache/syncope/fit/core/AuditITCase.java (hunk 13 above)
Comment (mmoayyed) [reply fit/core-reference/src/test/java/org/apache/syncope/fit/core/AuditITCase.java]: Done.

**Comment by mmoayyed** [comment]
Going to add a few more test cases based on Groups, etc but this should be essentially ready for further reviews.

File: core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java (hunk 3 above)
Comment (ilgrosso) [reply core/persistence-api/src/main/java/org/apache/syncope/core/persistence/api/dao/AuditDAO.java]: Given https://issues.apache.org/jira/browse/SYNCOPE-1517 I would simply leave things as they are ATM, refactoring will come later by thath.

**Comment by ilgrosso** [comment]
LGTM please @mmoayyed go ahead, squash and merge, thanks!

**Pull request merged**